go/storage/mkvs/db/badger/migrate.go

```diff
@@ -0,0 +1,693 @@
+	v4.readTxn = v4.db.db.NewTransactionAt(maxTimestamp, false)
+	defer func() {
+		// readTxn will change throughout the process, don't
+		// bind the defer to a particular instance.
+		v4.readTxn.Discard()
+	}()
+	v4.changeBatch = v4.db.db.NewWriteBatchAt(maxTimestamp)
+	defer func() {
+		// changeBatch will change throughout the process, don't
+		// bind the defer to a particular instance.
+		v4.changeBatch.Cancel()
+	}()
+
+	// Load migration metadata and set up saving on function return.
+	err := v4.meta.load(v4.db.db)
+	if err != nil && err != badger.ErrKeyNotFound {
+		return 0, err
+	}
+	v4.meta.BaseDBVersion = 3
+
+	// Count keys first, so we can report some sensible progress to the user.
+	// Badger says this should be fast.
+	if !v4.meta.InitComplete {
+		v4.helper.DisplayStepBegin("scanning database")
+		v4.meta.MetaCount = 0
+		func() {
+			opts := badger.DefaultIteratorOptions
+			opts.PrefetchValues = false
+			it := v4.readTxn.NewIterator(opts)
+			defer it.Close()
+			it.Rewind()
+			it.Seek([]byte{v3WriteLogKeyFmt.Prefix()})
+			for ; it.Valid(); it.Next() {
+				prefix := it.Item().Key()[0]
+				if prefix == v3MultipartRestoreNodeLogKeyFmt.Prefix() {
+					v4.meta.MultipartActive = true
+				}
+				if prefix != v3NodeKeyFmt.Prefix() && prefix != v3MultipartRestoreNodeLogKeyFmt.Prefix() {
+					v4.meta.MetaCount++
+				}
+			}
+		}()
+		v4.meta.InitComplete = true
+		v4.helper.DisplayStepEnd(fmt.Sprintf("%v keys to migrate", v4.meta.MetaCount))
+	}
+
+	// Migrate!
+
+	if !v4.meta.MetaComplete {
+		if err := v4.migrateMeta(); err != nil {
+			return 0, err
+		}
+		v4.meta.MetaComplete = true
+		v4.meta.LastKey = []byte{}
+		if err := v4.flush(true); err != nil {
+			return 0, fmt.Errorf("error migrating metadata: %w", err)
+		}
+	}
+
+	// All done, flush and clean up. The metadata blob will be removed
+	// in the defer handler.
+	v4.done = true
+	if err := v4.flush(true); err != nil {
+		return 0, err
+	}
+	return 4, nil
+}
+
+func Migrate(cfg *api.Config, helper MigrationHelper) (uint64, error) {
+	db := &badgerNodeDB{
+		logger:           logging.GetLogger("mkvs/db/badger/migrate"),
+		namespace:        cfg.Namespace,
+		discardWriteLogs: cfg.DiscardWriteLogs,
+	}
+	opts := commonConfigToBadgerOptions(cfg, db)
+
+	var err error
+	if db.db, err = badger.OpenManaged(opts); err != nil {
+		return 0, fmt.Errorf("mkvs/badger/migrate: failed to open database: %w", err)
+	}
+	defer db.Close()
+
+	// Make sure that we can discard any deleted/invalid metadata.
+	db.db.SetDiscardTs(tsMetadata)
+
+	// Load metadata.
+	lastVersion, err := func() (uint64, error) {
+		tx := db.db.NewTransactionAt(tsMetadata, false)
+		defer tx.Discard()
+
+		var migMeta migrationCommonMeta
+		item, rerr := tx.Get(migrationMetaKeyFmt.Encode())
+		if rerr == nil {
+			rerr = item.Value(func(data []byte) error {
+				return cbor.UnmarshalTrusted(data, &migMeta)
+			})
+			if rerr != nil {
+				return 0, rerr
```

Comment:
Wrap error using `fmt.Errorf` to provide more context of what failed. Also elsewhere.